File: 0141-linked-list-cycle/0141-linked-list-cycle.java
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) {
 *         val = x;
 *         next = null;
 *     }
 * }
 */
public class Solution {
    public boolean hasCycle(ListNode head) {
        ListNode fast = head;
        ListNode slow = head;
        while(fast!=null&&fast.next!=null){
            fast = fast.next.next;  // Move fast two steps
            slow = slow.next;       // Move slow one step
            if(fast==slow){        // Cycle detected
                return true;
            }
        }
      return false;
        
    }
}

File: 0141-linked-list-cycle/0141-linked-list-cycle.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def hasCycle(self, h: Optional[ListNode]) -> bool:
        if h is None:
            return False
        
        s, f = h, h.next

        while s and f:
            if not f or not f.next:
                return False
            
            if s == f:
                return True

            s = s.next
            f = f.next.next
        
        return False